src/celeritas/em/interactor/detail/IoniFinalStateHelper.hh

```diff
@@ -0,0 +1,128 @@
+
+  public:
+    // Construct from data
+    inline CELER_FUNCTION IoniFinalStateHelper(Energy inc_energy,
+                                               Real3 const& inc_direction,
+                                               Momentum inc_momentum,
+                                               Mass inc_mass,
+                                               Energy electron_energy,
+                                               Mass electron_mass,
+                                               ParticleId electron_id,
+                                               Secondary* secondary);
+
+    // Update the final state of the incident particle and secondary electron
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+  private:
+    // Incident particle energy [MeV]
+    real_type inc_energy_;
+    // Incident particle direction
+    Real3 const& inc_direction_;
+    // Incident particle momentum [MeV / c]
+    real_type inc_momentum_;
+    // Incident particle mass
+    real_type inc_mass_;
+    // Secondary electron energy [MeV]
+    real_type electron_energy_;
+    // Electron mass
+    real_type electron_mass_;
+    // Secondary electron ID
+    ParticleId electron_id_;
+    // Allocated secondary electron
+    Secondary* secondary_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from incident particle and exiting gamma data.
+ */
+CELER_FUNCTION
+IoniFinalStateHelper::IoniFinalStateHelper(Energy inc_energy,
+                                           Real3 const& inc_direction,
+                                           Momentum inc_momentum,
+                                           Mass inc_mass,
+                                           Energy electron_energy,
+                                           Mass electron_mass,
+                                           ParticleId electron_id,
+                                           Secondary* secondary)
+    : inc_energy_(value_as<Energy>(inc_energy))
+    , inc_direction_(inc_direction)
+    , inc_momentum_(value_as<Momentum>(inc_momentum))
+    , inc_mass_(value_as<Mass>(inc_mass))
+    , electron_energy_(value_as<Energy>(electron_energy))
+    , electron_mass_(value_as<Mass>(electron_mass))
+    , electron_id_(electron_id)
+    , secondary_{secondary}
+{
+    CELER_EXPECT(secondary_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Update the final state of the incident particle and secondary electron.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction IoniFinalStateHelper::operator()(Engine& rng)
+{
+    // TODO: If the \c UseAngularGeneratorFlag is set (false by default), use
+    // the angular generator interface to sample the secondary direction (see
+    // \c G4EmParameters::UseAngularGeneratorForIonisation())
```

Comment:
It seems like the flag in Geant4 basically lets users directly sample `secondary_->direction`... are we really ever going to support that? It would essentially require templating everything up through the model.

Reply:
Yeah almost certainly not :)